### PR TITLE
More new visual scripting features

### PR DESCRIPTION
**Pull request by C-Core**
* Hashed string and enums support
* Local variables
* Message handler and sender
* Reflected properties
* Exposed debug rendering
* Unit tests

With that the new visual scripting has basically the same features as the old visual scripting system.

### CODE REVIEW

File: Code/Engine/Core/Messages/Implementation/EventMessage.cpp [hunk 1]
```diff
@@ -7,6 +7,14 @@
...
 EZ_IMPLEMENT_MESSAGE_TYPE(ezEventMessage);
 EZ_BEGIN_DYNAMIC_REFLECTED_TYPE(ezEventMessage, 1, ezRTTIDefaultAllocator<ezEventMessage>)
+{
+  /*EZ_BEGIN_PROPERTIES
+  {
+    EZ_MEMBER_PROPERTY_READ_ONLY("SenderObject", m_hSenderObject),
+    EZ_MEMBER_PROPERTY_READ_ONLY("SenderComponent", m_hSenderComponent)
+  }
+  EZ_END_PROPERTIES;*/
+}
```
Comment (jankrassnigg) [inline Code/Engine/Core/Messages/Implementation/EventMessage.cpp]: Remove commented out code again?

File: Code/Engine/Core/World/GameObject.h [hunk 2]
```diff
@@ -424,14 +424,14 @@ class EZ_CORE_DLL ezGameObject final
   /// \param senderComponent The component that triggered the event in the first place. May be nullptr.
   ///        If not null, this information is stored in \a msg as ezEventMessage::m_hSenderObject and ezEventMessage::m_hSenderComponent.
   ///        This information is used to pass through more contextual information for the event handler.
-  ///        For instance, a trigger would pass through which object entered the trigger.
```
Comment (jankrassnigg) [inline Code/Engine/Core/World/GameObject.h:427]: Why this change? For a trigger it is quite important to know WHO entered the trigger.

File: Code/Engine/Core/World/Implementation/GameObject.cpp [hunk 3]
```diff
@@ -145,9 +171,29 @@ void ezGameObject::Reflection_SetMode(ezObjectMode::Enum mode)
...
+
+void ezGameObject::Reflection_SetGlobalRotation(const ezQuat qRotation)
+{
+  SetGlobalRotation(qRotation);
+}
+
+void ezGameObject::Reflection_SetGlobalScaling(const ezVec3 vScaling)
+{
+  SetGlobalScaling(vScaling);
+}
```
Comment (jankrassnigg) [inline Code/Engine/Core/World/Implementation/GameObject.cpp]: Why do we need these ?
Also why are the parameters sometimes const& and sometimes not ?

File: Code/EditorPlugins/VisualScript/EditorPluginVisualScript/VisualScriptClassAsset/VisualScriptClassAsset.cpp [hunk 4]
```diff
@@ -77,7 +80,7 @@ ezTransformStatus ezVisualScriptClassAssetDocument::InternalTransformAsset(ezStr
   }
 
   ezStringBuilder sDumpPath;
-  if (false)
+  if (GetProperties()->m_bDumpAST)
```
Comment (jankrassnigg) [inline Code/EditorPlugins/VisualScript/EditorPluginVisualScript/VisualScriptClassAsset/VisualScriptClassAsset.cpp:83]: This looks hacky

File: Code/Engine/Core/World/Implementation/GameObject.cpp (hunk 3 above)
Comment (C-Core) [reply Code/Engine/Core/World/Implementation/GameObject.cpp]: Because we have two overloads of SetGlobal... one with regular math classes, one with simd math classes. This doesn't work with the reflection since it doesn't know which one to pick. The missing & is wrong of course.

File: Code/Engine/Core/World/GameObject.h (hunk 2 above)
Comment (C-Core) [reply Code/Engine/Core/World/GameObject.h:427]: I only fixed the comment to state the truth. The trigger message has an extra member who entered the trigger.

File: Code/EditorPlugins/VisualScript/EditorPluginVisualScript/VisualScriptGraph/VisualScriptTypeDeduction.cpp [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    {
+      pinDataType = static_cast<const ezVisualScriptPin&>(connections[0]->GetSourcePin()).GetResolvedScriptDataType();
+    }
+    else
+    {
+      ezVariant var = pObject->GetTypeAccessor().GetValue(pPin->GetName());
+      pinDataType = ezVisualScriptDataType::FromVariantType(var.GetType());
+    }
+
+    deductedType = ezMath::Max(deductedType, pinDataType);
```
Comment (jankrassnigg) [inline Code/EditorPlugins/VisualScript/EditorPluginVisualScript/VisualScriptGraph/VisualScriptTypeDeduction.cpp:56]: Is it commented at ezVisualScriptDataType that the order is important ?
Maybe even add some static_asserts to prevent future changes.

File: Code/EditorPlugins/VisualScript/EditorPluginVisualScript/VisualScriptGraph/VisualScriptTypeDeduction.cpp [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+  const ezString& sTypeName = typeVar.Get<ezString>();
+  if (sTypeName.IsEmpty())
+    return nullptr;
+
+  const ezRTTI* pType = ezRTTI::FindTypeByName(sTypeName);
+  if (pType == nullptr)
+  {
+    ezStringBuilder sFullTypeName;
+    sFullTypeName.Set("ez", typeVar.Get<ezString>());
```
Comment (jankrassnigg) [inline Code/EditorPlugins/VisualScript/EditorPluginVisualScript/VisualScriptGraph/VisualScriptTypeDeduction.cpp:113]: nit: Could only do this, if it doesn't already start with "ez".

File: Code/EnginePlugins/VisualScriptPlugin/Runtime/VisualScript.h [hunk 7]
```diff
@@ -246,4 +259,20 @@ class EZ_VISUALSCRIPTPLUGIN_DLL ezVisualScriptExecutionContext
...
 
+struct ezVisualScriptSendMessageMode
+{
+  using StorageType = ezUInt8;
+
+  enum Enum
+  {
+    Direct,
+    Recursive,
+    Event,
```
Comment (jankrassnigg) [inline Code/EnginePlugins/VisualScriptPlugin/Runtime/VisualScript.h]: I assume this is mainly "internal" ? Or would it make sense to add some comment here ?

File: Code/EnginePlugins/VisualScriptPlugin/Runtime/VisualScriptNodeUserData.h [hunk 8]
```diff
@@ -74,7 +84,9 @@ namespace
     {
       EZ_SUCCEED_OR_RETURN(NodeUserData_Type::Serialize(nodeDesc, inout_stream, out_uiSize, out_uiAlignment));
 
-      inout_stream << nodeDesc.m_sTargetPropertyName;
+      const ezVariantArray& propertiesVar = nodeDesc.m_Value.Get<ezVariantArray>();
+
+      inout_stream << propertiesVar[0].Get<ezHashedString>();
```
Comment (jankrassnigg) [inline Code/EnginePlugins/VisualScriptPlugin/Runtime/VisualScriptNodeUserData.h:90]: I guess I don't understand this code. Why are we serializing only the first variable here ?

File: Code/EnginePlugins/VisualScriptPlugin/Runtime/VisualScriptNodeUserData.h [hunk 9]
```diff
@@ -123,17 +134,20 @@ namespace
...
       NodeUserData_Type::ToString(nodeDesc, out_sResult);
 
-      if (nodeDesc.m_sTargetPropertyName.IsEmpty() == false)
+      if (nodeDesc.m_Value.IsA<ezVariantArray>())
       {
-        out_sResult.Append(".", nodeDesc.m_sTargetPropertyName);
+        const ezVariantArray& propertiesVar = nodeDesc.m_Value.Get<ezVariantArray>();
+        if (propertiesVar.IsEmpty() == false)
+        {
+          out_sResult.Append(".", propertiesVar[0].Get<ezHashedString>());
```
Comment (jankrassnigg) [inline Code/EnginePlugins/VisualScriptPlugin/Runtime/VisualScriptNodeUserData.h:150]: Same question here. Why only look at item 0 ?

File: Code/EnginePlugins/VisualScriptPlugin/Runtime/VisualScriptNodeUserData.h [hunk 10]
```diff
@@ -142,13 +156,131 @@ namespace
 
   //////////////////////////////////////////////////////////////////////////
 
+  struct NodeUserData_TypeAndProperties : public NodeUserData_Type
+  {
+    ezUInt32 m_uiNumProperties;
```
Comment (jankrassnigg) [inline Code/EnginePlugins/VisualScriptPlugin/Runtime/VisualScriptNodeUserData.h]: = 0

File: Code/EnginePlugins/VisualScriptPlugin/Runtime/VisualScriptNodeUserData.h [hunk 11]
```diff
@@ -142,13 +156,131 @@ namespace
...
+
+#if EZ_ENABLED(EZ_PLATFORM_32BIT)
+    ezUInt32 m_uiPadding0;
+#endif
+
+    const ezAbstractProperty* m_Properties[1];
```
Comment (jankrassnigg) [inline Code/EnginePlugins/VisualScriptPlugin/Runtime/VisualScriptNodeUserData.h:169]: Okay... ?

File: Code/EnginePlugins/VisualScriptPlugin/Runtime/VisualScriptNodeUserData.h [hunk 12]
```diff
@@ -142,13 +156,131 @@ namespace
...
+      userData.m_uiNumProperties = uiCount;
+
+      ezHybridArray<const ezAbstractProperty*, 32> properties;
+      userData.m_pType->GetAllProperties(properties);
+
+      for (ezUInt32 i = 0; i < uiCount; ++i)
+      {
+        const ezAbstractProperty* pProperty = nullptr;
+        EZ_SUCCEED_OR_RETURN(NodeUserData_TypeAndProperty::ReadProperty(inout_stream, userData.m_pType, properties.GetArrayPtr(), pProperty));
+        userData.m_Properties[i] = pProperty;
```
Comment (jankrassnigg) [inline Code/EnginePlugins/VisualScriptPlugin/Runtime/VisualScriptNodeUserData.h:216]: Since m_Properties has a size of 1, shouldn't this at least assert something ?

File: Code/EnginePlugins/VisualScriptPlugin/Runtime/VisualScriptNodeUserData.h [hunk 13]
```diff
@@ -142,13 +156,131 @@ namespace
...
+      auto& userData = ref_node.InitUserData<NodeUserData_Switch>(inout_pAdditionalData, uiByteSize);
+      userData.m_uiNumCases = uiCount;
+
+      for (ezUInt32 i = 0; i < uiCount; ++i)
+      {
+        inout_stream >> userData.m_Cases[i];
+      }
+
+      return EZ_SUCCESS;
+    }
```
Comment (jankrassnigg) [inline Code/EnginePlugins/VisualScriptPlugin/Runtime/VisualScriptNodeUserData.h]: Same comments as for user data above

File: Code/EnginePlugins/VisualScriptPlugin/Runtime/VisualScriptNodeUserData.h [hunk 14]
```diff
@@ -142,13 +156,131 @@ namespace
...
+
+    static void ToString(const ezVisualScriptNodeDescription& nodeDesc, ezStringBuilder& out_sResult)
+    {
+    }
```
Comment (jankrassnigg) [inline Code/EnginePlugins/VisualScriptPlugin/Runtime/VisualScriptNodeUserData.h:277]: Missing impl ?

**Review by jankrassnigg — APPROVED** [review]

File: Code/EditorPlugins/VisualScript/EditorPluginVisualScript/VisualScriptClassAsset/VisualScriptClassAsset.cpp (hunk 4 above)
Comment (C-Core) [reply Code/EditorPlugins/VisualScript/EditorPluginVisualScript/VisualScriptClassAsset/VisualScriptClassAsset.cpp:83]: I know, haven't found a better solution yet. Can be cleaned up at some point.

File: Code/EnginePlugins/VisualScriptPlugin/Runtime/VisualScriptNodeUserData.h (hunk 8 above)
Comment (C-Core) [reply Code/EnginePlugins/VisualScriptPlugin/Runtime/VisualScriptNodeUserData.h:90]: Because for nodes with this user data there is only one "target" property. Other nodes can have multiple, like e.g. for send and receive message nodes this list would contain all "exposable" properties of the message type.

File: Code/EditorPlugins/VisualScript/EditorPluginVisualScript/VisualScriptGraph/VisualScriptTypeDeduction.cpp (hunk 5 above)
Comment (C-Core) [reply Code/EditorPlugins/VisualScript/EditorPluginVisualScript/VisualScriptGraph/VisualScriptTypeDeduction.cpp:56]: Added a comment

File: Code/EnginePlugins/VisualScriptPlugin/Runtime/VisualScriptNodeUserData.h (hunk 11 above)
Comment (C-Core) [reply Code/EnginePlugins/VisualScriptPlugin/Runtime/VisualScriptNodeUserData.h:169]: As discussed offline, also added a comment
